## <center>Universidad de las Fuerzas Armadas ESPE</center>


## <center>Actividad de Aprendizaje 3</center>

**Apellidos:** Flores Medrano

**Nombres:** Pamela Lissette

**Asignatura:** Métodos Numéricos

**NRC:** 3657

### Ejercicios


**1.  Para el sistema lineal:**

*$6x_1 = 12$*

*$6x_2+3x_1=-12$*

*$7x_3-2x_2+4x_1=14$*

*$21x_4+9x_3-3x_2+5x_1=-2$*

**Encuentre los valores de $x_1, x_2, x_3, x_4$ utilizando sustitución progresiva. Encuentre la descomposición LU del sistema anterior. ¿Qué puede concluir?**


**- Sustitución Progresiva**

_Ecuación 1_

$6x_1 = 12$

> $x_1 = 2$ 

_Ecuación 1 en 2_

$6x_2 + 3(2)= -12$

> $x_2= -3$

_Ecuación 1 y 2 en 3_

$7x_3 - 2(-3) + 4(2)= 14$

> $x_3= 0$

_Ecuación 1,2,3 en 4_

$21x_4 + 9(0) + -3(-3) + 5(2)= -2$

> $x_4= -1$

**DESCOMPOSICIÓN LU**

$\left(\begin{array}{cccc} {6} & {0} & {0} & {0} & \mbox{|} \ 12 \\ 3 & {6} & {0} & {0} & \mbox{|}\ -12\\ 4 & -2 & 7 & 0 & \mbox{|}\ 14 \\ 5 & -3 & 9 & 21 & \mbox{|}\ -2 \end{array}\right)$

$Ax=b$

$\begin{pmatrix}6&0&0&0\\ 3&6&0&0\\ 4&-2&7&0\\ 5&-3&9&21\end{pmatrix}\begin{pmatrix}x_1\\ x_2\\ x_3\\ x_4\end{pmatrix}=\begin{pmatrix}12\\ -12\\ 14\\ -2\end{pmatrix}$

$A=LU$

$\begin{pmatrix}6&0&0&0\\ 3&6&0&0\\ 4&-2&7&0\\ 5&-3&9&21\end{pmatrix}=\begin{pmatrix}1&0&0&0\\ l_{2,1}&1&0&0\\ l_{3,1}&l_{3,2}&1&0\\ l_{4,1}&l_{4,2}&l_{4,3}&1\end{pmatrix}\begin{pmatrix}u_{1,1}&u_{1,2}&u_{1,3}&u_{1,4}\\ 0&u_{2,2}&u_{2,3}&u_{2,4}\\ 0&0&u_{3,3}&u_{3,4}\\ 0&0&0&u_{4,4}\end{pmatrix}$

$=\begin{pmatrix}u_{1,1}&u_{1,2}&u_{1,3}&u_{1,4}\\ \:l_{2,1}\cdot \:u_{1,1}&l_{2,1}\cdot \:u_{1,2}+u_{2,2}&l_{2,1}\cdot \:u_{1,3}+u_{2,3}&l_{2,1}\cdot \:u_{1,4}+u_{2,4}\\ \:l_{3,1}\cdot \:u_{1,1}&l_{3,1}\cdot \:u_{1,2}+l_{3,2}\cdot u_{2,2}&l_{3,1}\cdot \:u_{1,3}+l_{3,2}\cdot \:u_{2,3}+u_{3,3}&l_{3,1}\cdot \:u_{1,4}+l_{3,2}\cdot \:u_{2,4}+u_{3,4}\\ \:l_{4,1}\cdot \:u_{1,1}&l_{4,1}\cdot \:u_{1,2}+l_{4,2}\cdot \:u_{2,2}&l_{4,1}\cdot \:u_{1,3}+l_{4,2}\cdot \:u_{2,3}+l_{4,3}\cdot \:u_{3,3}&l_{4,1}\cdot \:u_{1,4}+l_{4,2}\cdot \:u_{2,4}+l_{4,3}\cdot \:u_{3,4}+u_{4,4}\end{pmatrix}$

**Resolución:**

$u_{1,\:1}=6\:;\:u_{1,\:2}=0\:;\:u_{1,\:3}=0\:;\:u_{1,\:4}=0$

• $\frac{1}{2}\left(0\right)+u_{2,2}=6$

> $u_{2,2}=6$

$\frac{1}{2}\left(0\right)+u_{2,3}=0$

> $u_{2,3}=0$

$\frac{1}{2}\left(0\right)+u_{2,4}=0$

> $u_{2,4}=0$

• $\frac{2}{3}\left(0\right)+\left(-\frac{1}{3}\right)\left(0\right)+u_{3,3}=7$

> $u_{3,3}=7$

• $\frac{2}{3}\left(0\right)+\left(-\frac{1}{3}\right)\left(0\right)+u_{3,4}=0$

> $u_{3,4}=0$

• $\frac{5}{6}\left(0\right)+\left(-\frac{1}{2}\right)\left(0\right)+\frac{9}{7}\left(0\right)+u_{4,4}=21$

> $u_{4,4}=21$

• $l_{2,1}\left(6\right)=3\:$

> $l_{2,1}=\frac{1}{2}$

• $l_{3,1}\left(6\right)=4\:$

> $l_{3,1}=\frac{2}{3}$

• $\frac{2}{3}\left(0\right)+l_{3,2}\left(6\right)=-2$

> $l_{3,2}=-\frac{1}{3}$

• $l_{4,1}\left(6\right)=5$

> $l_{4,1}=\frac{5}{6}$

• $\frac{5}{6}\left(0\right)+l_{4,2}\left(6\right)=-3$

> $l_{4,2}=-\frac{1}{2}$

• $\frac{5}{6}\left(0\right)+\left(-\frac{1}{2}\right)\left(0\right)+l_{4,3}\left(7\right)=9$

> $l_{4,3}=\frac{9}{7}$

**Por lo tanto:**

$L=\begin{pmatrix}1&0&0&0\\ \frac{1}{2}&1&0&0\\ \frac{2}{3}&-\frac{1}{3}&1&0\\ \frac{5}{6}&-\frac{1}{2}&\frac{9}{7}&1\end{pmatrix};\:U=\begin{pmatrix}6&0&0&0\\ 0&6&0&0\\ 0&0&7&0\\ 0&0&0&21\end{pmatrix}$

**Sabiendo que Ax=b; A=LU entonces [L][Ux]=[B]**

$L\left[U\begin{pmatrix}x_1\\ x_2\\ x_3\\ x_4\end{pmatrix}\right]=\begin{pmatrix}12\\ -12\\ 14\\ -2\end{pmatrix}$

**Si Ux=[y] entonces L[y]=b**

$L\left[\begin{pmatrix}y_1\\ y_2\\ y_3\\ y_4\end{pmatrix}\right]=\begin{pmatrix}12\\ -12\\ 14\\ -2\end{pmatrix}$

$\begin{pmatrix}1&0&0&0\\ \:\frac{1}{2}&1&0&0\\ \:\frac{2}{3}&-\frac{1}{3}&1&0\\ \:\frac{5}{6}&-\frac{1}{2}&\frac{9}{7}&1\end{pmatrix}\begin{pmatrix}y_1\\ y_2\\ y_3\\ y_4\end{pmatrix}=\begin{pmatrix}12\\ -12\\ 14\\ -2\end{pmatrix}$

**Sustitución Progresiva**

_Ecuación 1_

> $y_1= 12$

_Ecuación 1 en 2_

$\frac{1}{2}y_1+y_2=-12$

$\frac{1}{2}\left(12\right)+y_2=-12$

> $y_2=-18$

_Ecuación 1 y 2 en 3_

$\frac{2}{3}y_1-\frac{1}{3}y_2+y_3= 14$

$\frac{2}{3}\left(12\right)-\frac{1}{3}\left(-18\right)+y_3= 14$

> $y_3= 0$

_Ecuación 1,2,3 en 4_

$\frac{5}{6}y_1-\frac{1}{2}y_2+\frac{9}{7}y_3+y_4=-2$

$\frac{5}{6}\left(12\right)-\frac{1}{2}\left(-18\right)+\frac{9}{7}\left(0\right)+y_4=-2$

> $y_4=-21$

**[Ux]=y**

$U\begin{pmatrix}x_1\\ x_2\\ x_3\\ x_4\end{pmatrix}=\begin{pmatrix}12\\ -18\\ 0\\ -21\end{pmatrix}$

$\begin{pmatrix}6&0&0&0\\ 0&6&0&0\\ 0&0&7&0\\ 0&0&0&21\end{pmatrix}\begin{pmatrix}x_1\\ x_2\\ x_3\\ x_4\end{pmatrix}=\begin{pmatrix}12\\ -18\\ 0\\ -21\end{pmatrix}$

**Sustitución Regresiva**

_Ecuación 1_

$6x_1= 12$

> $x_1= 2$

_Ecuación 2_

$6x_2= -18$

> $x_2= -3$

_Ecuación 3_

$7x_3= 0$

> $x_3= 0$

_Ecuación 4_

$21x_4= -21$

> $x_4= -1$


**Conclusión**

Llegamos a la misma respuesta tanto utilizando la descomposición LU y sustitución progresiva. 

Observamos que la descomposición LU con respecto a la eliminación gaussiana es más eficiente pues se realiza menos operaciones en menos tiempo si tomamos en cuenta su programación.


**2. Resuelva las siguientes matrices usando la descomposición de Cholesky**

<center>$A=\begin{pmatrix}4&6&10\\ 6&25&19\\ 10&19&62\end{pmatrix}$
    
**_Primer Paso:_ Verificar que es simétrica y positiva.**
    
$\:A^T=\begin{pmatrix}4&6&10\\ \:6&25&19\\ \:10&19&62\end{pmatrix}^T=\begin{pmatrix}4&6&10\\ \:6&25&19\\ \:10&19&62\end{pmatrix}$

> Es simétrica.
    
Menores Principales:

$M_1=\begin{pmatrix}4\end{pmatrix}$

$det M_1 = 4$
    
$M_2=\begin{pmatrix}4&6\\ 6&25\end{pmatrix}$

$det M_2 = 100-36 = 64$

$M_3=\begin{pmatrix}4&6&10\\ 6&25&19\\ 10&19&62\end{pmatrix}$

$det M_3= \left(4\cdot 25\cdot 62\right)+\left(6\cdot 19\cdot 10\right)+\left(6\cdot 19\cdot 10\right)-\left[\left(25\cdot 10\cdot 10\right)+\left(19\cdot 19\cdot 4\right)+\left(6\cdot 6\cdot 62\right)\right]$
    
$det M_3= 6200+1140+1140-2500-1444-2232= 2304$
    
> Es positiva.

**DESCOMPOSICIÓN CHOLESKY**

$Ax=b$

$A=L\cdot L^T$
    
$A=\begin{pmatrix}4&6&10\\ 6&25&19\\ 10&19&62\end{pmatrix}=\begin{pmatrix}l_{1,1}&0&0\\ l_{2,1}&l_{2,2}&0\\ l_{3,1}&l_{3,2}&l_{3,3}\end{pmatrix}\begin{pmatrix}l_{1,1}&1_{2,1}&l_{3,1}\\ 0&l_{2,2}&l_{3,2}\\ 0&0&l_{3,3}\end{pmatrix}$

$=\begin{pmatrix}l^2_{1,1}&l_{1,1}\cdot l_{2,1}&l_{1,1}\cdot l_{3,1}\\ l_{2,1}\cdot l_{1,1}&l^2_{2,1}\cdot l^2_{2,2}&l_{2,1}\cdot l_{3,1}+l_{2,2}\cdot l_{3,2}\\ l_{3,1}\cdot l_{1,1}&l_{3,1}\cdot l_{2,1}+l_{3,2}\cdot l_{2,2}&l^2_{3,1}+l^2_{3,2}+l^2_{3,3}\end{pmatrix}$
    
Algoritmo de Cholesky
    
$l_{k,i}=\frac{\left[a_{k,i}-\left(\sum _{j=1}^{i-1}\:\:l_{i,j}\cdot l_{k,j}\right)\right]}{l_{i,i}}$
    
$l_{k,k}=\sqrt{a_{k,k}-\sum _{j=1}^{k-1}\:l^2_{k,j}\:}$
    
**Resolución**

Cuando k=1
    
$l_{1,1}=\sqrt{a_{1,1}}=\sqrt{4}=2$

Cuando k=2
    
$l_{2,1}=\frac{a_{2,1}}{l_{1,1}}=\frac{6}{2}=3$
    
$l_{2,2}=\sqrt{a_{2,2}-l^2_{2,1}}=\sqrt{25-3^2}=4$
    
Cuando k=3
    
$l_{3,1}=\frac{a_{3,1}}{l_{1,1}}=\frac{10}{2}=5$
    
$l_{3,2}=\frac{a_{3,2}-l_{2,1}\cdot l_{3,1}}{l_{2,2}}=\frac{19-3\left(5\right)}{4}=1$
    
$l_{3,3}=\sqrt{a_{3,3}-\left(l^2_{3,1}+l^2_{3,2}\right)}=\sqrt{62-\left(25+1\right)}=6$

Por lo tanto:
    
$L=\begin{pmatrix}2&0&0\\ 3&4&0\\ 5&1&6\end{pmatrix}$

Usando descomposición de Cholesky, tenemos que: 
$A=L\cdot L^T$
   
**Respuesta**
    
$A=\begin{pmatrix}2&0&0\\ 3&4&0\\ 5&1&6\end{pmatrix}\begin{pmatrix}2&3&5\\ 0&4&1\\ 0&0&6\end{pmatrix}$


-------------------------------------------------------------------------
<center>$B=\begin{pmatrix}4&6&10\\ 6&3&19\\ 10&19&62\end{pmatrix}$

**_Primer Paso:_ Verificar que es simétrica y positiva.**

$B^T=\begin{pmatrix}4&6&10\\ \:\:6&3&19\\ \:\:10&19&62\end{pmatrix}^T=\begin{pmatrix}4&6&10\\ \:\:\:6&3&19\\ \:\:\:10&19&62\end{pmatrix}$
    
> Es simétrica.
    
Menores Principales:
    
$M_1=\begin{pmatrix}4\end{pmatrix}$

$det M_1 = 4$
    
$M_2=\begin{pmatrix}4&6\\ 6&3\end{pmatrix}$

$det M_2 = 12-36 = -24$
    
$M_3=\begin{pmatrix}4&6&10\\ 6&3&19\\ 10&19&62\end{pmatrix}$

$det M_3= \left(4\cdot 3\cdot 62\right)+\left(6\cdot 19\cdot 10\right)+\left(6\cdot 19\cdot 10\right)-\left[\left(3\cdot 10\cdot 10\right)+\left(19\cdot 19\cdot 4\right)+\left(6\cdot 6\cdot 62\right)\right]$
    
$det M_3=-952$
    
> NO es positiva
    
**Respuesta:**
    
No podemos realizar la descomposición por Cholesky pues aunque la matriz es simétrica no cumple con que ésta sea positiva y consecuentemente tendremos raíces imaginarias.

**3. Utilice el método de Jacobi, Gauss-Seidel y SOR (ω = 1.1) para resolver el siguiente sistema lineal con una precisión de cuatro cifras decimales.**


<center>$\begin{pmatrix}7&1&-1&2\\ 1&8&0&-2\\ -1&0&4&-1\\ 2&-2&-1&6\end{pmatrix}\begin{pmatrix}x_1\\ x_2\\ x_3\\ x_4\end{pmatrix}=\begin{pmatrix}3\\ -5\\ 4\\ -3\end{pmatrix}$
    
**Compare el número de iteraciones necesario en cada algoritmo. Analice el error cometido si la solución exacta es $x\:=\:\left(1,\:−1,\:1,\:−1\right)^T$**


**MÉTODO DE JACOBI**

In [46]:
import numpy as np

A = np.array([[ 7, 1,-1,  2], 
              [ 1, 8, 0, -2], 
              [-1, 0, 4, -1],
              [ 2,-2, -1, 6]])

b = np.array([3, -5, 4, -3])

x0 = np.zeros(len(b))

#Comprobamos que la matriz de entrada es diagonal dominante
def diag_dom(X):
    
    #Encontramos los coeficientes de la diagonal
    D = np.diag(np.abs(X))
    S = np.sum(np.abs(X), axis=1) - D 
    
    if np.all(D > S):
        print("La matriz SI es diagonal dominante\n")
    else:
        print("La matriz NO es diagonal dominante\n")
    return

diag_dom(A)

#Implementamos la función para el método de Jacobi
def Jacobi(A, b, x0, kmax = 100, tol = 1e-4):
    #Utilizamos el método copy para que la lista original no cambie cuando se modifique la nueva lista en el método.
    x_new = x0.copy() 
    #Usamos la función numpy.shape() el cual encuentra la longitud de las dimensiones del array A.
    fil, col = np.shape(A)
    
    #Método de Jacobi
    for k in range(kmax): 
        x_copy = x_new.copy()
        for i in range(fil):     
            sum = 0        
            for j in range(col):
                if j != i:
                    #Fórmula de iteración
                    sum += A[i, j] * x_new[j]
            x_copy[i] = (b[i] - sum) / A[i][i]
        #Diferencia con Gauss-Seidel
        x_new = x_copy.copy()
        
        # Criterio de parada
        Ax = np.dot(A, x_new)               
        cp = np.linalg.norm(b - Ax)
        print("Iteración", k+1,"°",'Residuo: {0:10.6g}'.format(cp))
        if cp < tol:
            print ("\nEl número máximo de iteraciones alcanzado es: ", k+1, " iteraciones")
            break
    return x_new

#Imprimimos resultados
x = Jacobi(A, b, x0)   
print("\nLa solución por el Método de Jacobi es:\n", x)


La matriz SI es diagonal dominante

Iteración 1 ° Residuo:    3.18784
Iteración 2 ° Residuo:    1.46156
Iteración 3 ° Residuo:   0.751959
Iteración 4 ° Residuo:   0.357989
Iteración 5 ° Residuo:    0.18525
Iteración 6 ° Residuo:  0.0882664
Iteración 7 ° Residuo:  0.0456901
Iteración 8 ° Residuo:   0.021771
Iteración 9 ° Residuo:  0.0112699
Iteración 10 ° Residuo: 0.00537005
Iteración 11 ° Residuo: 0.00277984
Iteración 12 ° Residuo: 0.00132458
Iteración 13 ° Residuo: 0.000685678
Iteración 14 ° Residuo: 0.000326723
Iteración 15 ° Residuo: 0.00016913
Iteración 16 ° Residuo: 8.05899e-05

El número máximo de iteraciones alcanzado es:  16  iteraciones

La solución por el Método de Jacobi es:
 [ 0.99998943 -0.99998943  1.00000058 -0.999982  ]


**MÉTODO GAUSS-SEIDEL**

In [37]:
import numpy as np

A = np.array([[ 7, 1,-1,  2], 
              [ 1, 8, 0, -2], 
              [-1, 0, 4, -1],
              [ 2,-2, -1, 6]])
                  
b = np.array([3, -5, 4, -3])

x0 = np.zeros(len(b)) 

#Comprobamos que la matriz de entrada es diagonal dominante
def diag_dom(X):
    
    #Encontramos los coeficientes de la diagonal
    D = np.diag(np.abs(X))
    S = np.sum(np.abs(X), axis=1) - D 
    
    if np.all(D > S):
        print("La matriz SI es diagonal dominante\n")
    else:
        print("La matriz NO es diagonal dominante\n")
    return 

diag_dom(A)

def GaussSeidel(A, b, x0, kmax = 100, tol = 1e-4):
    #Utilizamos el método copy para que la lista original no cambie cuando se modifique la nueva lista en el método.
    x_new = x0.copy()
    #Usamos la función numpy.shape() el cual encuentra la longitud de las dimensiones del array A.
    fil, col = np.shape(A)
    
    #Método de Gauss-Seidel
    for k in range(kmax): 
        for i in range(fil):     
            sum = 0        
            for j in range(col):
                if j != i:
                    #Fórmula de iteración
                    sum += A[i, j] * x_new[j]
            x_new[i] = (b[i] - sum) / A[i][i]
            
 # Criterio de parada
        Ax = np.dot(A, x_new)               
        cp = np.linalg.norm(b - Ax)
        print("Iteración",k+1,"°",'Residuo: {0:10.6g}'.format(cp))
        if cp < tol:
            print ("\nEl número máximo de iteraciones alcanzado es: ", k+1, " iteraciones")
            break
        
    return x_new

#Imprimimos resultados
x = GaussSeidel(A, b, x0)
print("\nLa solución por el Método de Gauss Seidel es:\n", x)

La matriz SI es diagonal dominante

Iteración 1 ° Residuo:    3.50648
Iteración 2 ° Residuo:   0.829666
Iteración 3 ° Residuo:    0.20442
Iteración 4 ° Residuo:  0.0504185
Iteración 5 ° Residuo:  0.0124362
Iteración 6 ° Residuo: 0.00306752
Iteración 7 ° Residuo: 0.000756639
Iteración 8 ° Residuo: 0.000186634
Iteración 9 ° Residuo: 4.60353e-05

El número máximo de iteraciones alcanzado es:  9  iteraciones

La solución por el Método de Gauss Seidel es:
 [ 0.99999343 -0.99999468  1.00000286 -0.99999556]


**MÉTODO SUCCESSIVE OVER RELAXATION (SOR)**

In [40]:
import numpy as np

A = np.array([[ 7, 1,-1,  2], 
              [ 1, 8, 0, -2], 
              [-1, 0, 4, -1],
              [ 2,-2, -1, 6]])
                  
b = np.array([3, -5, 4, -3])

x0 = np.zeros(len(b)) 

#Comprobamos que la matriz de entrada es diagonal dominante
def diag_dom(X):
    
    #Encontramos los coeficientes de la diagonal
    D = np.diag(np.abs(X))
    S = np.sum(np.abs(X), axis=1) - D 
    
    if np.all(D > S):
        print("La matriz SI es diagonal dominante\n")
    else:
        print("La matriz NO es diagonal dominante\n")
    return

diag_dom(A)

def SOR(A, b, x0, omega=1.1, kmax = 100, tol = 1e-4):
    #Utilizamos el método copy para que la lista original no cambie cuando se modifique la nueva lista en el método.
    x_new = x0.copy()
    #Usamos la función numpy.shape() el cual encuentra la longitud de las dimensiones del array A.
    fil, col = np.shape(A)
    
    #Método SOR
    for k in range(kmax): 
        for i in range(fil):     
            sum = 0        
            for j in range(col):
                if j != i:
                    #Fórmula de iteración
                    sum += A[i, j] * x_new[j]
            x_new[i] = (1 - omega) * x_new[i] + (omega / A[i][i]) * (b[i] - sum)
            
# Criterio de parada
        Ax = np.dot(A, x_new)               
        cp = np.linalg.norm(b - Ax)
        print("Iteración",k+1,"°",'Residuo: {0:10.6g}'.format(cp))
        if cp < tol:
            print ("\nEl número máximo de iteraciones alcanzado es: ", k+1, " iteraciones")
            break
        
    return x_new

x = SOR(A, b, x0)
    
print("\nLa solución por el Método de SOR es:\n", x)

La matriz SI es diagonal dominante

Iteración 1 ° Residuo:    3.61793
Iteración 2 ° Residuo:   0.371416
Iteración 3 ° Residuo:   0.019736
Iteración 4 ° Residuo:  0.0039075
Iteración 5 ° Residuo: 0.000318519
Iteración 6 ° Residuo: 9.82736e-06

El número máximo de iteraciones alcanzado es:  6  iteraciones

La solución por el Método de SOR es:
 [ 0.99999859 -0.99999975  1.0000006  -0.99999833]


**ANÁLISIS**

**_Número de Iteraciones de cada algoritmo_**

Método de Jacobi: **16** iteraciones.

Método de Gauss-Seidel: **9** iteraciones.

Método SOR: **6** iteraciones.

Como podemos observar el método que menos iteraciones tuvo tomando en cuenta la precisión pedida fue el método de SOR, consecuentemente el método de Gauss-Seidel y por último el método de Jacobi.

**_Error cometido si la solución exacta es $x\:=\:\left(1,\:−1,\:1,\:−1\right)^T$_**

Usaremos el error absoluto siendo x aproximada el vector solución de cada método y x exacta el vector dado en el enunciado.

**Solución Método de Jacobi:**$\begin{pmatrix}0.99998943&-0.99998943&1.00000058&-0.999982\end{pmatrix}^T$


$e_{abs}$= $x-x\:̃$ = $\begin{pmatrix}0.00001057&0.00001057&0.00000058&0.000018\end{pmatrix}^T$


In [21]:
import numpy as np

x = [[0.00001057, 0.00001057, 0.00000058, 0.000018]]

print("El error en el método de Jacobbi es:",np.linalg.norm(x))

El error en el método de Jacobbi es: 2.3404832834267372e-05


**Solución Método de Gauss-Seidel:**$\begin{pmatrix}0.99999343&-0.99999468&1.00000286&-0.99999556\end{pmatrix}^T$

$e_{abs}$ = $x-x\:̃$ = $\begin{pmatrix}0.00000657&0.00000532&0.00000286&0.00000444\end{pmatrix}^T$

In [22]:
import numpy as np

x = [[0.00000657, 0.00000532, 0.00000286, 0.00000444]]

print("El error en el método de Gauss-Seidel es:",np.linalg.norm(x))

El error en el método de Gauss-Seidel es: 9.967973715856197e-06


**Solución Método SOR:**$\begin{pmatrix}0.99999859&-0.99999975&1.0000006&-0.99999833\end{pmatrix}^T$

$e_{abs}$ = $x-x\:̃$ = $\begin{pmatrix}0.00000141&0.00000025&0.0000006&0.00000167\end{pmatrix}^T$

In [23]:
import numpy as np

x = [[0.00000141, 0.00000025, 0.0000006, 0.00000167]]

print("El error en el método SOR es:",np.linalg.norm(x))

El error en el método SOR es: 2.2802412153103456e-06


**Conclusión**

Como conclusión podemos observar que el método que cometió menor error es el método SOR.

**4. Programe un algoritmo para encontrar la norma de Frobenius para una matriz cuadrada de cualquier dimensión.**


In [35]:
import numpy as np

#Pedimos por pantalla la dimensión de la matriz
print("Introduzca la dimensión de la matriz cuadrada:")
n = int(input())

#Creamos una matriz cuadrada randómica
A = np.random.random((n,n))

#Algoritmo Norma de Frobenius
def norma_frobenius(A):
    sum = 0
    for i in A:
        for elem in i:
            sum = sum + elem**2
    sum = sum**(1 / 2)
    print("La matriz es:\n", A)
    return sum

print("\nLa norma de Frobenius para la matriz cuadrada dada es: ",norma_frobenius(A))

Introduzca la dimensión de la matriz cuadrada:
3
La matriz es:
 [[0.43346695 0.2564235  0.78155124]
 [0.10322777 0.06974845 0.94092327]
 [0.7923981  0.93186866 0.65449073]]

La norma de Frobenius para la matriz cuadrada dada es:  1.9209264546990528
